### PR TITLE
Bug 1132014 - Reading list home panel

### PR DESCRIPTION
**Pull request by st3fan**
This patch adds support for a new Reading List Home Panel:

* Adds temporary code where a long press on the bookmark star adds the page to your reading list
* Adds a reading list home panel
* Adds opening reading list items from the reading list

Split in two separate commits.

### CODE REVIEW

File: Client/Frontend/Reader/ReaderMode.swift [hunk 1]
```diff
@@ -100,6 +110,8 @@ class ReaderMode: BrowserHelper, ReaderModeStyleViewControllerDelegate {
     var state: ReaderModeState = ReaderModeState.Unavailable
     private var originalURL: NSURL?
 
+    var activateImmediately: Bool = false
```
Comment (thebnich) [inline Client/Frontend/Reader/ReaderMode.swift]: Nit: <code>: Bool</code> unnecessary.

File: Client/Frontend/Home/ReaderPanel.swift [hunk 2]
```diff
@@ -3,51 +3,52 @@
...
+class ReaderPanel: UITableViewController, HomePanel {
 
-    private var sites = [Site]()
-    /*
-        Site(title:"Royals Sweep Orioles to Advance to World Series", url:"http://www.nytimes.com/2014/10/16/sports/baseball/royals-keep-rolling-and-advance-to-the-world-series.html?hp&action=click&pgtype=Homepage&version=LargeMediaHeadlineSum&module=photo-spot-region&region=top-news&WT.nav=top-news&_r=0"),
-        Site(title:"How Not to Be Fooled by Odds", url:"http://www.nytimes.com/2014/10/16/upshot/how-not-to-be-fooled-by-odds.html?hp&action=click&pgtype=Homepage&version=HpSum&module=second-column-region&region=top-news&WT.nav=top-news"),
-        Site(title:"Against Rules, Amber Vinson, Dallas Worker With Ebola, Boarded Plane", url:"http://www.nytimes.com/2014/10/16/us/ebola-outbreak-texas.html?hp&action=click&pgtype=Homepage&version=LedeSum&module=first-column-region&region=top-news&WT.nav=top-news")
-    ]
-    */
+    var readingListItemsCursor: Cursor? = nil
```
Comment (wesj) [inline Client/Frontend/Home/ReaderPanel.swift]: You could make this non-optional and instead just set it to an empty cursor when its not assigned. Those can even have some "state", so we could do things like "Loading".

File: Client/Frontend/Reader/ReaderMode.js [hunk 3]
```diff
@@ -14,7 +14,8 @@ var _firefox_ReaderMode = {
             // Short circuit in case we already ran Readability. This mostly happens when going
             // back/forward: the page will be cached and the result will still be there.
             if (_firefox_ReaderMode.readabilityResult && _firefox_ReaderMode.readabilityResult.content) {
-                webkit.messageHandlers.readerModeMessageHandler.postMessage("Available");
+                console.log({Type: "ReaderModeStateChange", Value: "Available"});
```
Comment (wesj) [inline Client/Frontend/Reader/ReaderMode.js]: These are leaked to webdevs. Be careful not to leave them in in releases...

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 4]
```diff
@@ -149,10 +149,25 @@ class BrowserViewController: UIViewController {
...
-        tabManager.selectedTab?.loadRequest(NSURLRequest(URL: url))
+
+        if let tab = tabManager.selectedTab {
+            if ReaderMode.isReaderModeURL(url) {
+                if let readerMode = tab.getHelper(name: "ReaderMode") as? ReaderMode {
+                    // Switch to reader mode immediately when we detect it can be activated. The reader mode will still
+                    // call its delegate to let us know its state changed so that we can update the UI.
+                    readerMode.activateImmediately = true
+                    // We don't show the initial page when opening a reader: url. This will probably change to some overlay on top of the webview.
+                    tab.hideContent(animated: false)
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: This isn't great since webViews are stacked, meaning going into reader mode actually shows the tab behind the current one. Sounds like you're planning on changing this, so this is fine for now, but please file a follow-up so this doesn't get forgotten.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 5]
```diff
@@ -315,6 +330,17 @@ extension BrowserViewController: BrowserToolbarDelegate {
         }
     }
 
+    // TODO: This is temporary way to add items to your reading list until we have actual buttons
+    func browserToolbarDidLongPressBookmark(browserToolbar: BrowserToolbar) {
+        if let tab = tabManager.selectedTab? {
+            if let url = tab.url?.absoluteString {
+                profile.readingList.add(item: ReadingListItem(url: url, title: tab.title)) { (success) -> Void in
+                    // Nothing to do here
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Can't you just pass <code>nil</code> instead of making an empty callback?

File: Client/Frontend/Reader/ReaderMode.swift [hunk 6]
```diff
@@ -128,23 +140,41 @@ class ReaderMode: BrowserHelper, ReaderModeStyleViewControllerDelegate {
...
-    private func constructAboutReaderURL(originalURL: NSURL?) -> NSURL {
-        if let url = originalURL?.absoluteString {
-            if let encoded = url.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.alphanumericCharacterSet()) {
-                if let aboutReaderURL = NSURL(string: "about:reader?url=\(encoded)") {
-                    return aboutReaderURL
+    private func handleReaderModeStateChange(state: ReaderModeState) {
+        self.state = state
+        delegate?.readerMode(self, didChangeReaderModeState: state, forBrowser: browser!)
+
+        if activateImmediately && state == ReaderModeState.Available {
```
Comment (thebnich) [inline Client/Frontend/Reader/ReaderMode.swift]: This condition seems broken. What if the user clicks something in the reading list that isn't actually readable? For exmaple, pages that are readable on Android but not mobile, pages that require sign-in, clicking a reading list entry while the user is behind a captive portal, etc. Then reader mode won't be enabled, and we'd end up auto-entering reader mode the next time the user _does_ encounter a readable page, which is broken.

One fix would be to remove <code>&& state == ReaderModeState.Available</code> so that <code>activateImmediately</code> will be set to <code>false</code> unconditionally. Note that <code>enableReaderMode()</code> already checks the reader mode state, so checking <code>ReaderModeState.Available</code> here is redundant. Maybe we should rename the method to <code>enableReaderModeIfAvailable()</code>?

File: Client/Frontend/Reader/ReaderMode.swift [hunk 7]
```diff
@@ -5,6 +5,15 @@
 import Foundation
 import WebKit
 
+enum ReaderModeMessageType: String {
+    case StateChange = "ReaderModeStateChange"
+    case PageEvent = "ReaderPageEvent"
+}
+
+enum ReaderPageEvent: String {
+    case PageShow = "PageShow"
```
Comment (thebnich) [inline Client/Frontend/Reader/ReaderMode.swift]: Are you planning on including additional events here? If not, having an enum for a single value seems odd.

File: Client/Frontend/Home/ReaderPanel.swift [hunk 8]
```diff
@@ -3,51 +3,52 @@
...
     }
 
+    // MARK: UITableViewDataSourceDelegate
+
     override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
-        return sites.count
+        if let cursor = self.readingListItemsCursor {
+            return cursor.count
+        }
+        return 0
```
Comment (thebnich) [inline Client/Frontend/Home/ReaderPanel.swift]: Suggestion: shorten this to <code>return self.readingListItemsCursor?.count ?? 0</code>.

**Comment by thebnich** [comment]
Looks good as an initial implementation. r=me with comments addressed.

Rather than hiding the webView, loading the page, then showing the webView again, another option could be to load the reader URL/do the parsing in a separate (hidden) webView. This might give us a bit more flexibility with transitions since we could show the previous page longer, or show some loading animation directly in the about:reader HTML while we wait for the result.
